ai_edge_torch/generative/examples/stable_diffusion/decoder.py

```diff
@@ -15,99 +15,251 @@
+                    norm_1="15.groupnorm_1",
+                    norm_2="15.groupnorm_2",
+                    conv_1="15.conv_1",
+                    conv_2="15.conv_2",
+                    residual_layer="15.residual_layer",
+                ),
+                autoencoder_loader.ResidualBlockTensorNames(
+                    norm_1="16.groupnorm_1",
+                    norm_2="16.groupnorm_2",
+                    conv_1="16.conv_1",
+                    conv_2="16.conv_2",
+                ),
+                autoencoder_loader.ResidualBlockTensorNames(
+                    norm_1="17.groupnorm_1",
+                    norm_2="17.groupnorm_2",
+                    conv_1="17.conv_1",
+                    conv_2="17.conv_2",
+                ),
+            ],
+            upsample_conv="19",
+        ),
+        autoencoder_loader.UpDecoderBlockTensorNames(
+            residual_block_tensor_names=[
+                autoencoder_loader.ResidualBlockTensorNames(
+                    norm_1="20.groupnorm_1",
+                    norm_2="20.groupnorm_2",
+                    conv_1="20.conv_1",
+                    conv_2="20.conv_2",
+                    residual_layer="20.residual_layer",
+                ),
+                autoencoder_loader.ResidualBlockTensorNames(
+                    norm_1="21.groupnorm_1",
+                    norm_2="21.groupnorm_2",
+                    conv_1="21.conv_1",
+                    conv_2="21.conv_2",
+                ),
+                autoencoder_loader.ResidualBlockTensorNames(
+                    norm_1="22.groupnorm_1",
+                    norm_2="22.groupnorm_2",
+                    conv_1="22.conv_1",
+                    conv_2="22.conv_2",
+                ),
+            ],
+        ),
+    ],
+    final_norm="23",
+    conv_out="25",
+)
 
-class AttentionBlock(nn.Module):
 
-  def __init__(self, channels):
-    super().__init__()
-    self.groupnorm = nn.GroupNorm(32, channels)
-    self.attention = SelfAttention(1, channels)
-
-  def forward(self, x):
-    residue = x
-    x = self.groupnorm(x)
-
-    n, c, h, w = x.shape
-    x = x.view((n, c, h * w))
-    x = x.transpose(-1, -2)
-    x = self.attention(x)
-    x = x.transpose(-1, -2)
-    x = x.view((n, c, h, w))
-
-    x += residue
-    return x
+class Decoder(nn.Module):
 
-
-class ResidualBlock(nn.Module):
-
-  def __init__(self, in_channels, out_channels):
+  def __init__(self, config: unet_cfg.AutoEncoderConfig):
     super().__init__()
-    self.groupnorm_1 = nn.GroupNorm(32, in_channels)
-    self.conv_1 = nn.Conv2d(in_channels, out_channels, kernel_size=3, padding=1)
-
-    self.groupnorm_2 = nn.GroupNorm(32, out_channels)
-    self.conv_2 = nn.Conv2d(out_channels, out_channels, kernel_size=3, padding=1)
-
-    if in_channels == out_channels:
-      self.residual_layer = nn.Identity()
-    else:
-      self.residual_layer = nn.Conv2d(
-          in_channels, out_channels, kernel_size=1, padding=0
+    self.config = config
```

Comment:
as discussed we can add some high-level drawing of model structure in comment section to help understand the model definition.

Reply:
Done